### PR TITLE
[SPARK-20544][SPARKR] R wrapper for input_file_name

### PR DESCRIPTION
**Pull request by zero323**
## What changes were proposed in this pull request?

Adds wrapper for `o.a.s.sql.functions.input_file_name`

## How was this patch tested?

Existing unit tests, additional unit tests, `check-cran.sh`.

### CODE REVIEW

**Comment by felixcheung** [comment]
hmm, not clear why AppVeyor failed. you could trigger it again by closing and re-opening this PR

File: R/pkg/R/functions.R [hunk 1]
```diff
@@ -3890,3 +3890,23 @@ setMethod("not",
             jc <- callJStatic("org.apache.spark.sql.functions", "not", x@jc)
             column(jc)
           })
+
+#' input_file_name
+#'
+#' Creates a string column for the file name of the current Spark task.
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: I actually find this description in Scala API quite a bit confusing - what is "Spark task" and how it has "file name"?

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 2]
```diff
@@ -1656,6 +1656,18 @@ test_that("greatest() and least() on a DataFrame", {
   expect_equal(collect(select(df, least(df$a, df$b)))[, 1], c(1, 3))
 })
 
+test_that("input_file_name()", {
+  path <- tempfile(pattern = "input_file_name_test", fileext = ".txt")
+  write.table(iris[1:50, ], path, row.names = FALSE, col.names = FALSE)
+
+  df <- read.text(path)
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: does it work with `df <- read.json(jsonPath)`?

if yes, consider adding to the test for column functions

(again, this is regarding: https://github.com/apache/spark/pull/17817 to consolidate/skip tests)

**Comment by zero323** [comment]
> hmm, not clear why AppVeyor failed. you could trigger it again by closing and re-opening this PR
without affecting Jenkins

Look I'll have to rebase it anyway but thank you so much for the hint. I've been meaning to ask if there is some equivalent of Jenkins helpers.

File: R/pkg/inst/tests/testthat/test_sparkSQL.R (hunk 2 above)
Comment (zero323) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Should work with any file input as far as I remember. I'd skip collecting but there have been some issues with PySpark in the past.

**Review by felixcheung — APPROVED** [review]
LGTM

File: R/pkg/R/functions.R [hunk 3]
```diff
@@ -3974,3 +3974,23 @@ setMethod("grouping_id",
...
+#'
+#' @rdname input_file_name
+#' @name input_file_name
+#' @aliases input_file_name,missing-method
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: actually, could you add `@family normal_funcs` here? I missed this earlier and in the other PR.

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/R/functions.R (hunk 3 above)
Comment (zero323) [inline R/pkg/R/functions.R]: Done.

File: R/pkg/R/functions.R (hunk 1 above)
Comment (zero323) [inline R/pkg/R/functions.R]: How about the new one?

File: R/pkg/R/functions.R [hunk 4]
```diff
@@ -3974,3 +3974,24 @@ setMethod("grouping_id",
             jc <- callJStatic("org.apache.spark.sql.functions", "grouping_id", jcols)
             column(jc)
           })
+
+#' input_file_name
+#'
+#' Creates a string column with the input file name for a given row
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: this actually makes a lot more sense...

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
merged to master

**Pull request closed without merging**